doc/libspi.2
.\" man page for Munts Technologies Linux Simple I/O Library
.\"
.\" Copyright (C)2016-2025, Philip Munts dba Munts Technologies.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\" * Redistributions of source code must retain the above copyright notice,
.\"   this list of conditions and the following disclaimer.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.TH libspi 2 "25 September 2025" "version 1" "Linux Simple I/O Library"
.SH NAME
.B libspi
\-\- Linux Simple I/O Library: SPI Module
.SH SYNOPSIS
.nf
.B #include <libsimpleio/libspi.h>

.BI "void SPI_open(const char *" name ", int32_t " mode ", int32_t " wordsize ","
.BI "  int32_t " speed ", int32_t *" fd ", int32_t *" error ");"

.BI "void SPI_close(int32_t " fd ", int32_t *" error ");"

.BI "void SPI_transaction(int32_t " spifd ", int32_t " csfd ", void *" cmd ","
.BI "  int32_t " cmdlen ", int32_t " delayus ", void *" resp ", int32_t " resplen ","
.BI "  int32_t *" error ");"

.fi
Link with
.BR -lsimpleio .
.SH DESCRIPTION
.nh
All functions return either
.B 0
(upon success) or an
.B errno
value (upon failure) in
.IR *error .
.PP
.B SPI_open()
opens an SPI slave device.  The device name,
.BR /dev/spidevx.x ,
must be passed in the
.IR name
parameter. The
.I mode
parameter specifies the SPI bus transfer mode,
.BR 0 " to " 3 .
The
.I wordsize
parameter specifies the SPI bus transfer unit size, usually
.BR 8 ", " 16 ", or " 32 " bits."
Some SPI controllers only allow 8-bit transfers.
The
.I speed
parameter specifies the SPI bus transfer speed in bits per second.
Upon success, a file descriptor for the
SPI slave device is returned in
.IR *fd .
.PP
.B SPI_close()
closes a previously opened SPI slave device.
.PP
.B SPI_transaction()
performs a single SPI bus transaction, with optional transmit and receive
phases. Either the address of a command message and its length must be passed in the
.IR cmd " and " cmdlen
parameters, or
.BR NULL " and " 0
for a receive only transaction.  The
.I delayus
parameter indicates the time in microseconds between the transmit and receive phases.  It
should be set long enough for the SPI slave device to execute the command and generate its
response.
Either the address of a receive buffer and its size must be passed in the
.IR resp " and " resplen
parameters, or
.BR NULL " and " 0
for a transmit only transaction.  The
.IR csfd
parameter should be set to
.BR SPI_CS_AUTO
to use the hardware controlled slave chip select signal or set to the open file descriptor
for a GPIO pin to use for the software controlled slave chip select signal.
.SH SEE ALSO
.BR libsimpleio "(2), " libadc "(2), " libdac "(2), " libevent "(2), " libgpio "(2),"
.br
.BR libhidraw "(2), " libiio "(2), " libipv4 "(2), " libi2c "(2), " liblinux "(2),"
.br
.BR liblinx "(2), " libpwm "(2), " libserial "(2), " libstream "(2), " libwatchdog "(2)"
.SH AUTHOR
Philip Munts dba Munts Technologies.


csharp/libSPI.cs
// C# binding for SPI device services in libsimpleio.so

// Copyright (C)2017, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

using System.Runtime.InteropServices;

namespace libsimpleio
{
    /// <summary>
    /// Wrapper for libsimpleio SPI device services.
    /// </summary>
    public class libSPI
    {
        /// <summary>
        /// Open a Linux SPI device.
        /// </summary>
        /// <param name="devname">Device node name.</param>
        /// <param name="mode">SPI transfer mode (0 .. 3)</param>
        /// <param name="wordsize">SPI transfer word size (8, 16, or 32).</param>
        /// <param name="speed">SPI transfer speed in Hz.</param>
        /// <param name="fd">File descriptor.</param>
        /// <param name="error">Error code.  Zero upon success or an <code>errno</code>
        /// value upon failure.</param>
        ///<remarks>The Linux kernel create a device nodes for each SPI slave
        ///device, of the form
        ///<code>/dev/spidevX.Y</code> where <code>X</code> is the SPI bus
        ///controller number and <code>Y</code> is the SPI slave select number.</remarks>
        [DllImport("simpleio")]
        public static extern void SPI_open(string devname, int mode,
            int wordsize, int speed, out int fd, out int error);

        /// <summary>
        /// Send bytes to and/or receive bytes from a Linux SPI device.
        /// </summary>
        /// <param name="fd">File descriptor.</param>
        /// <param name="csfd">Chip select file descriptor.</param>
        /// <param name="cmd">Source buffer.</param>
        /// <param name="cmdlen">Source buffer size.</param>
        /// <param name="delayus">Delay in microseconds between the write and read operations.</param>
        /// <param name="resp">Destination buffer.</param>
        /// <param name="resplen">Destination buffer size.</param>
        /// <param name="error">Error code.  Zero upon success or an <code>errno</code>
        /// value upon failure.</param>
        [DllImport("simpleio")]
        public static extern void SPI_transaction(int fd, int csfd, byte[] cmd,
            int cmdlen, int delayus, byte[] resp, int resplen, out int error);

        /// <summary>
        /// Close a Linux SPI device.
        /// </summary>
        /// <param name="fd">File descriptor.</param>
        /// <param name="error">Error code.  Zero upon success or an <code>errno</code>
        /// value upon failure.</param>
        [DllImport("simpleio")]
        public static extern void SPI_close(int fd, out int error);
    }
}
